cdc/sink/common/flow_control.go

```diff
@@ -0,0 +1,209 @@
+	if nBytes >= c.Quota {
+		return cerrors.ErrFlowControllerEventLargerThanQuota.GenWithStackByArgs(nBytes, c.Quota)
+	}
+
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	calledBack := false
+	for {
+		if atomic.LoadUint32(&c.IsAborted) == 1 {
+			return cerrors.ErrFlowControllerAborted.GenWithStackByArgs()
+		}
+		if c.Consumed+nBytes < c.Quota {
+			break
+		}
+
+		if !calledBack {
+			calledBack = true
+			err := blockCallBack()
+			if err != nil {
+				return errors.Trace(err)
+			}
+		}
+		c.cond.Wait()
+	}
+
+	c.Consumed += nBytes
+	return nil
+}
+
+// ForceConsume is called when blocking is not acceptable and the limit can be violated
+// for the sake of avoid deadlock. It merely records the increased memory consumption.
+func (c *TableMemorySizeController) ForceConsume(nBytes uint64) error {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	if atomic.LoadUint32(&c.IsAborted) == 1 {
+		return cerrors.ErrFlowControllerAborted.GenWithStackByArgs()
+	}
+
+	c.Consumed += nBytes
+	return nil
+}
+
+// Release is called when a chuck of memory is done being used.
+func (c *TableMemorySizeController) Release(nBytes uint64) {
+	c.mu.Lock()
+
+	if c.Consumed < nBytes {
+		defer c.mu.Unlock()
+		log.Panic("TableMemorySizeController: releasing more than consumed, report a bug",
+			zap.Uint64("consumed", c.Consumed),
+			zap.Uint64("released", nBytes))
+	}
+
+	c.Consumed -= nBytes
+	if c.Consumed < c.Quota {
+		c.mu.Unlock()
+		c.cond.Signal()
+		return
+	}
+
+	c.mu.Unlock()
+}
+
+// Abort interrupts any ongoing ConsumeWithBlocking call
+func (c *TableMemorySizeController) Abort() {
+	atomic.StoreUint32(&c.IsAborted, 1)
+	c.cond.Signal()
+}
+
+// TableFlowController provides a convenient interface to control the memory consumption of a per table event stream
+type TableFlowController struct {
+	memoryController *TableMemorySizeController
+
+	mu    sync.Mutex
+	queue deque.Deque
+
+	lastCommitTs uint64
+}
+
+type commitTsSizeEntry struct {
+	CommitTs uint64
+	Size     uint64
+}
+
+// NewTableFlowController creates a new TableFlowController
+func NewTableFlowController(quota uint64) *TableFlowController {
+	return &TableFlowController{
+		memoryController: NewTableMemorySizeController(quota),
+		queue:            deque.NewDeque(),
+	}
+}
+
+// Consume is called when an event has arrived for being processed by the sink.
+// It will handle transaction boundaries automatically, and will not block intra-transaction.
+func (c *TableFlowController) Consume(commitTs uint64, size uint64, blockCallBack func() error) error {
+	lastCommitTs := atomic.LoadUint64(&c.lastCommitTs)
+
+	if commitTs < lastCommitTs {
+		log.Panic("commitTs regressed, report a bug",
+			zap.Uint64("commitTs", commitTs),
+			zap.Uint64("lastCommitTs", c.lastCommitTs))
+	}
+
+	if commitTs > lastCommitTs {
+		atomic.StoreUint64(&c.lastCommitTs, commitTs)
+		err := c.memoryController.ConsumeWithBlocking(size, blockCallBack)
```

Comment:
Memory allocation is one by one, could this becomes bottleneck, is there any plan to introduce batch consume mechanism or pre-memory allocation?
Besides we should add some benchmark for the flow controller in later PRs.

Reply:
This is not a bottleneck. I have pinpointed the bottleneck to the sink manager.